### PR TITLE
channelz: cleanup channel registration if Dial fails

### PR DESCRIPTION
**Pull request by lyuxuan**

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -153,6 +153,18 @@ func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *
 			})
 		}
 		cc.csMgr.channelzID = cc.channelzID
+		defer func() {
```
Comment (menghanl) [inline clientconn.go]: cc.Close() will also remove channel and add trace channel deleted.

Move `RegisterChannel` down (right before defer with `cc.Close()`), then this should be covered by cc.Close().

File: clientconn.go [hunk 2]
```diff
@@ -196,18 +208,6 @@ func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *
 		defer cancel()
```
Comment (daviddrysdale) [inline clientconn.go]: Does this work if there's a timeout set (`cc.dopts.timeout > 0`) ?  In that case, the `defer`red actions will happen in the reverse order with this PR, so the auto-`cancel()` of the replaced `ctx` (with a timeout) happens before the check whether `ctx` is `Done()`, which means a cancellation error will always be returned.

Comment (menghanl) [reply clientconn.go]: Right!
I filed #2736 to track. Will have a solution shortly!
Thanks for noting and mentioning this!!

Comment (daviddrysdale) [reply clientconn.go]: I've got a fix & testcase on its way.

**Pull request merged**